File: src/main/java/com/stage/teamb/services/publication/PublicationService.java
package com.stage.teamb.services.publication;

import com.stage.teamb.dtos.employee.EmployeeDTO;
import com.stage.teamb.dtos.publication.PublicationCreateDTO;
import com.stage.teamb.dtos.publication.PublicationDTO;
import com.stage.teamb.dtos.publication.PublicationGetDTO;
import com.stage.teamb.dtos.rating.RatingDTO;
import com.stage.teamb.models.Publication;

import java.util.List;
import java.util.Optional;

public interface PublicationService {


    List<PublicationGetDTO> findAllPublications(Long authId);

    PublicationGetDTO findPublicationById(Long id, Long authUserId);

    List<PublicationGetDTO> findAllByEmployeeId(Long employeeId);

    PublicationGetDTO createPublication(PublicationCreateDTO publicationDTO, Long id);

    PublicationDTO updatePublication(Long publicationId, PublicationDTO publicationDTO);

    void deletePublication(Long publicationId);

    EmployeeDTO findEmployeeByPublicationId(Long publicationId);

    PublicationGetDTO associateEmployeeWithPublication(Long publicationId, Long employeeId);

    PublicationGetDTO disassociateEmployeeFromPublication(Long publicationId);


    List<RatingDTO> getRatingForPublication(Long publicationId);

    List<Publication> findAll();

    Optional<Publication> findOne(Long id);

    Publication saveOne(Publication Published);

    void deleteOne(Long id);

    Boolean existsById(Long id);
}

File: src/main/resources/teamb-front/src/app/shared/services/publication.service.ts
import { Injectable, OnDestroy } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, Subscription } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class PublicationService  {
  
  private baseUrl = 'http://localhost:8082/api/publications';

  constructor(private http: HttpClient) {}

  findAllPublications(): Observable<any[]> {
    const request = this.http.get<any[]>(`${this.baseUrl}/find/all`);
    return request;
  }

  findPublicationById(id: number): Observable<any> {
    const request = this.http.get<any>(`${this.baseUrl}/get/${id}`);  
    return request;
  }

  findAllByEmployeeId(employeeId: number): Observable<any[]> {
    const request = this.http.get<any[]>(`${this.baseUrl}/employee/${employeeId}`);
    return request;
  }

  createPublication(publicationDTO: any): Observable<any> {
    const request = this.http.post<any>(`${this.baseUrl}/create`, publicationDTO);
    return request;
  }

  updatePublication(publicationId: number, publicationDTO: any): Observable<any> {
    const request = this.http.put<any>(`${this.baseUrl}/update/${publicationId}`, publicationDTO);    
    return request;
  }

  deletePublication(publicationId: number): Observable<void> {
    const request = this.http.delete<void>(`${this.baseUrl}/delete/${publicationId}`);
    return request;
  }

  findEmployeeByPublicationId(publicationId: number): Observable<any> {
    const request = this.http.get<any>(`${this.baseUrl}/${publicationId}/employee`);
    return request;
  }

  associateEmployeeWithPublication(publicationId: number, employeeId: number): Observable<any> {
    const request = this.http.post<any>(`${this.baseUrl}/${publicationId}/employee/${employeeId}`, {});
    return request;
  }

  disassociateEmployeeFromPublication(publicationId: number): Observable<void> {
    const request = this.http.delete<void>(`${this.baseUrl}/${publicationId}/employee`);
    return request;
  }

  getRatingForPublication(publicationId: number): Observable<any[]> {
    const request = this.http.get<any[]>(`${this.baseUrl}/${publicationId}/rating`);
    return request;
  }

  
}
